### PR TITLE
🚧 collections

### PR DESCRIPTION
**Pull request by herrmannplatz**

### CODE REVIEW

File: test/unit_tests/message/MessageCategorizationSpec.coffee [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see http://www.gnu.org/licenses/.
+#
+
+# grunt test_init && grunt test_run:message/MessageCategorization
+
+describe 'z.message.MessageCategorization.category_from_event', ->
```
Comment (bennycode) [inline test/unit_tests/message/MessageCategorizationSpec.coffee]: We have the pattern of:

```coffee
describe 'z.message.MessageCategorization', ->
  describe 'category_from_event', ->
    it 'insures that events should have a category of type UNDEFINED', ->
      ...
```

File: app/script/view_model/content/CollectionViewModel.coffee [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+z.ViewModel.content ?= {}
+
+# Parent: z.ViewModel.ContentViewModel
+class z.ViewModel.content.CollectionViewModel
+  constructor: (element_id, @conversation_repository, @collection_details) ->
+    @logger = new z.util.Logger 'z.ViewModel.CollectionViewModel', z.config.LOGGER.OPTIONS
+
+    @conversation_et = ko.observable()
+
+    @images = ko.observableArray()
```
Comment (bennycode) [inline app/script/view_model/content/CollectionViewModel.coffee]: Can be sorted from A-Z:

```coffee
@audio = ko.observableArray()
@files = ko.observableArray()
@images = ko.observableArray()
@links = ko.observableArray()
@video = ko.observableArray()
```

File: app/script/view_model/ImageDetailViewViewModel.coffee [hunk 3]
```diff
@@ -22,52 +22,30 @@ z.ViewModel ?= {}
...
     @image_modal.destroy() if @image_modal?
-    @image_modal = new zeta.webapp.module.Modal '#detail-view', @_hide_callback, @_before_hide_callback
+    @image_modal = new zeta.webapp.module.Modal '#detail-view', undefined, @_before_hide_callback
     @image_modal.show()
 
-    @_show_image()
+    amplify.publish z.event.WebApp.ANALYTICS.EVENT, z.tracking.SessionEventName.INTEGER.IMAGE_DETAIL_VIEW_OPENED
+
+    @image_src src
+    setTimeout =>
```
Comment (bennycode) [inline app/script/view_model/ImageDetailViewViewModel.coffee]: `window.setTimeout`

File: app/script/main/app.coffee [hunk 4]
```diff
@@ -65,10 +65,14 @@ class z.main.App
     service.web_socket              = new z.event.WebSocketService @auth.client
 
     service.client                  = new z.client.ClientService @auth.client, service.storage
-    service.conversation            = new z.conversation.ConversationService @auth.client, service.storage
     service.notification            = new z.event.NotificationService @auth.client, service.storage
     service.announce                = new z.announce.AnnounceService()
 
+    if true # z.util.Environment.browser.edge
```
Comment (bennycode) [inline app/script/main/app.coffee]: So should this become `if z.util.Environment.browser.edge`?

File: app/script/entity/message/Message.coffee [hunk 5]
```diff
@@ -61,6 +61,9 @@ class z.entity.Message
     @visible = ko.observable true
     @version = 1
 
+    # z.message.MessageCategory
```
Comment (bennycode) [inline app/script/entity/message/Message.coffee:64]: Do we need this comment?

File: app/script/main/app.coffee (hunk 4 above)
Comment (herrmannplatz) [reply app/script/main/app.coffee]: yes, will revert it

File: app/script/entity/message/Message.coffee (hunk 5 above)
Comment (herrmannplatz) [reply app/script/entity/message/Message.coffee:64]: just to see the type

File: app/page/template/content/collection-details.htm [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  <!-- ko if: conversation_et() -->
+
+    <div class="content-titlebar">
+
+      <div class="content-titlebar-items-left">
+        <span class="content-titlebar-icon icon-back" data-bind="click: click_on_back_button"></span>
+      </div>
+
+      <span class="content-titlebar-items-center">
+        <span><span class="label-bold-xs">Pictures</span> <span class="text-graphite">in</span> <span class="text-graphite" data-bind="text: conversation_et().display_name()"></span></span>
```
Comment (gregor) [inline app/page/template/content/collection-details.htm:14]: Ensure strings are localizable and make use of the current template type for the correct naming.

File: app/page/template/content/collection-details.htm [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+      </span>
+
+    </div>
+
+    <div class="content-list">
+
+      <div class="collection-images">
+
+        <!-- ko if: template() === 'images' -->
```
Comment (gregor) [inline app/page/template/content/collection-details.htm:23]: Can we introduce a CollectionCategory or simular enum for these?

File: app/page/template/content/collection.htm [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+    <div class="content-list collection-list">
+
+      <!-- ko ifnot: no_items_found() -->
+
+        <!-- ko if: images().length -->
+          <section class="collection-section" data-uie-name="collection-section-image">
+            <header>
+              <span class="collection-header-icon icon-library"></span>
+              <span class="label-bold-xs">Pictures</span>
```
Comment (gregor) [inline app/page/template/content/collection.htm:25]: Ensure localizable strings

File: app/page/template/content/collection.htm [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+              <!-- /ko -->
+            </div>
+          </section>
+        <!-- /ko -->
+
+        <!-- ko if: links().length -->
+          <section class="collection-section" data-uie-name="collection-section-link">
+            <header>
+              <span class="collection-header-icon icon-link"></span>
+              <span class="label-bold-xs">Links</span>
```
Comment (gregor) [inline app/page/template/content/collection.htm:42]: Ensure localizable strings

File: app/page/template/content/collection.htm [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+              <!-- /ko -->
+            </div>
+          </section>
+        <!-- /ko -->
+
+        <!-- ko if: video().length -->
+          <section class="collection-section" data-uie-name="collection-section-video">
+            <header>
+              <span class="collection-header-icon icon-video"></span>
+              <span class="label-bold-xs">Videos</span>
```
Comment (gregor) [inline app/page/template/content/collection.htm:59]: Ensure localizable strings

File: app/page/template/content/collection.htm [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+              <!-- /ko -->
+            </div>
+          </section>
+        <!-- /ko -->
+
+        <!-- ko if: audio().length -->
+          <section class="collection-section" data-uie-name="collection-section-audio">
+            <header>
+              <span class="collection-header-icon icon-microphone"></span>
+              <span class="label-bold-xs">Audio messages</span>
```
Comment (gregor) [inline app/page/template/content/collection.htm:76]: Ensure localizable strings

File: app/page/template/content/collection.htm [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+              <!-- /ko -->
+            </div>
+          </section>
+        <!-- /ko -->
+
+        <!-- ko if: files().length -->
+          <section class="collection-section" data-uie-name="collection-section-file">
+            <header>
+              <span class="collection-header-icon icon-file"></span>
+              <span class="label-bold-xs">files</span>
```
Comment (gregor) [inline app/page/template/content/collection.htm:93]: Ensure localizable strings

File: app/page/template/content/collection.htm [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+          </section>
+        <!-- /ko -->
+
+      <!-- /ko -->
+
+      <!-- ko if: no_items_found() -->
+
+        <div class="absolute-center" data-uie-name="collection-no-items">
+          <div class="collection-no-items-icon icon-collection"></div>
+          <div class="label-bold-xs text-graphite text-center">No items</div>
```
Comment (gregor) [inline app/page/template/content/collection.htm:112]: Ensure localizable strings

File: app/script/conversation/ConversationRepository.coffee [hunk 14]
```diff
@@ -1945,6 +1957,8 @@ class z.conversation.ConversationRepository
   @return [Promise] Promise that resolves with the message entity for the event
   ###
   _add_event_to_conversation: (json, conversation_et) ->
+    message_et = @event_mapper.map_json_event json, conversation_et, true
+  _add_event_to_conversation: (json, conversation_et) ->
```
Comment (gregor) [inline app/script/conversation/ConversationRepository.coffee]: merge mess...

File: app/script/message/MessageCategorization.coffee [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see http://www.gnu.org/licenses/.
+#
+
+window.z ?= {}
+z.message ?= {}
+
+z.message.MessagaCategorization = do ->
```
Comment (gregor) [inline app/script/message/MessageCategorization.coffee]: an a wrong -> `z.message.MessageCategorization`

File: app/script/storage/StorageService.coffee [hunk 16]
```diff
@@ -170,6 +179,11 @@ class z.storage.StorageService
           if event.type is z.event.Client.CONVERSATION.DELETE_EVERYWHERE
             event.time = new Date(event.time).toISOString()
       @db.version(9).stores version_9
+      @db.version(10).stores version_10
+      .upgrade (transaction) =>
+        @logger.warn "Database upgrade to version #{@db.verno}", transaction
+        transaction[@OBJECT_STORE_CONVERSATION_EVENTS].toCollection().modify (event) ->
+          event.category = z.message.MessagaCategorization.category_from_event event
```
Comment (gregor) [inline app/script/storage/StorageService.coffee]: a -> e: here and 12 other spots

File: app/script/view_model/content/CollectionDetailsViewModel.coffee [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+class z.ViewModel.content.CollectionDetailsViewModel
+  constructor: (element_id) ->
+    @logger = new z.util.Logger 'z.ViewModel.CollectionDetailsViewModel', z.config.LOGGER.OPTIONS
+
+    @template = ko.observable()
+    @conversation_et = ko.observable()
+
+    @items = ko.observableArray()
+
+  set_conversation: (conversation_et, category, items) =>
```
Comment (gregor) [inline app/script/view_model/content/CollectionDetailsViewModel.coffee:33]: Naming is a little confusing. You do not set the conv but the collection category and items for a specific conv.

File: app/script/view_model/content/CollectionViewModel.coffee [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+        @populate_items message_ets
+
+  populate_items: (message_ets) =>
+    images = []
+    files = []
+    audio = []
+    video = []
+    links = []
+
+    for message_et, i in message_ets
```
Comment (gregor) [inline app/script/view_model/content/CollectionViewModel.coffee]: No need to keep track of the index if you do not use it.

**Pull request merged**